edgelet/edgelet-http/src/lib.rs

```diff
@@ -69,6 +77,102 @@ const TCP_SCHEME: &str = "tcp";
 #[cfg(target_os = "linux")]
 const FD_SCHEME: &str = "fd";
 
+#[derive(Clone)]
+pub struct PemCertificate {
+    cert: Vec<u8>,
+    key: Option<Vec<u8>>,
+    username: Option<String>,
+    password: Option<String>,
+}
+
+impl PemCertificate {
+    pub fn new(
+        cert: Vec<u8>,
+        key: Option<Vec<u8>>,
+        username: Option<String>,
+        password: Option<String>,
+    ) -> Self {
+        PemCertificate {
+            cert,
+            key,
+            username,
+            password,
+        }
+    }
+
+    pub fn get_certificate(&self) -> &[u8] {
+        &self.cert
+    }
+
+    pub fn from<C>(id_cert: &C) -> Result<Self, Error>
+    where
+        C: Certificate,
+    {
+        let cert = id_cert
+            .pem()
+            .map(|cert_buffer| cert_buffer.as_ref().to_owned())
+            .context(ErrorKind::IdentityCertificate)?;
+
+        let key = match id_cert.get_private_key() {
+            Ok(Some(PrivateKey::Ref(ref_))) => Some(ref_.into_bytes().clone()),
+            Ok(Some(PrivateKey::Key(KeyBytes::Pem(buffer)))) => Some(buffer.as_ref().to_vec()),
+            _ => None,
+        };
+
+        Ok(PemCertificate::new(cert, key, None, None))
+    }
+
+    pub fn get_identity(&self) -> Result<Identity, Error> {
+        let mut certs =
+            X509::stack_from_pem(&self.cert).with_context(|_| ErrorKind::IdentityCertificate)?;
```

Comment:
You only need `with_context` if the context requires computation that you would rather avoid if there is no error. Stuff like building a format string, etc.

For a simple enum variant like this, just `.context(ErrorKind::IdentityCertificate)` is good enough.

Same for the other uses of `with_context`

Reply:
Thanks good to know. I've addressed your comment here and everywhere else.